src/matchmaker.ts

```diff
@@ -1,44 +1,91 @@
-        this.intervalId = setInterval(() => { this.poll(); }, vars.SCHED_INTERVAL);
+        if (_.isNil(this.intervalId)) {
+            this.intervalId = setInterval(() => {
+                this.poll().catch((e) => { winston.error(e); });
+            }, vars.SCHED_INTERVAL);
+        }
     }
 
     stop() {
-        clearInterval(this.intervalId);
+        if (!_.isNil(this.intervalId)) {
+            clearInterval(this.intervalId);
+            this.intervalId = undefined;
+        }
     }
 
-    async getTeamsRandomOrder(): Promise<db.Team[]> {
-        return db.connection("teams").select("*")
-            .then(db.rowsToTeams)
-            .then((teams) => teams) // add permutation
+    async getPairedTeams() {
+        return db.connection.from(function (this: knex.QueryBuilder) {
+            this.from("submissions as subs")
+                .select("subs.id as subId", "subs.team_id as teamId", db.connection.raw("max(subs.version) as recent_version"))
+                .where({ status: "finished" })
+                .groupBy("subId")
+                .as("recent_subs");
+        }).join("submissions", function () {
+            this.on("recent_subs.teamId", "submissions.team_id").andOn("recent_subs.recent_version", "submissions.version");
+        })
+            .select("submissions.id as id", "recent_subs.teamId")
+            .then(permute)
+            .then(generateDerangedPairs)
+            .then((pairs): [IRecentSub, IRecentSub][] => pairs)
             .catch((e: Error) => { throw e; });
     }
 
     async scheduledNum(): Promise<number> {
-        return 4;
+        return db.connection("games").where({ status: "queued" }).count("*")
+            .then(([{ count }]): number => count)
+            .catch((e: Error) => { throw e; });
     }
 
     async poll(): Promise<void> {
-        let teams: db.Team[] = [];
-        try {
-            teams = await this.getTeamsRandomOrder();
-        } catch (e) {
-            winston.error(e);
-        }
+        if (await this.scheduledNum() < vars.SCHED_MAX) {
+            let pairs: Array<[IRecentSub, IRecentSub]> = [];
+            try {
+                pairs = await this.getPairedTeams();
+            } catch (e) {
+                winston.error(e);
+            }
 
-        teams.forEach(({ id, name }) => winston.info(`Team: ${id}, ${name}`));
+            pairs.forEach(([first, second]) => winston.info(`Matchup: (${first.teamId}, ${second.teamId})`));
+
+            for (const [{ id: firstId }, { id: secondId }] of pairs) {
+                for (let i = 0; i < vars.REPLICATIONS; i++) {
+                    const [{ id }] = await db.connection("games")
+                        .insert({ status: "queued" }, "*")
+                        .then(db.rowsToGames);
+                    await db.connection("games_submissions")
+                        .insert([{ game_id: id, submission_id: firstId }, { game_id: id, submission_id: secondId }])
+                        .then();
```

Comment:
Is this .then needed?

Reply:
Nope. I thought it was in order to kick off the query. Maybe it would be necessary without the await preceding it.